core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -1586,6 +1596,42 @@ trait ZIOFunctions extends Serializable {
    */
   final def descriptor: UIO[Fiber.Descriptor] = ZIO.Descriptor
 
+  /**
+   * Checks the interrupt status, and produces the effect returned by the
+   * specified callback.
+   */
+  final def checkInterruptible[R >: LowerR, E <: UpperE, A](f: Boolean => ZIO[R, E, A]): ZIO[R, E, A] =
+    new ZIO.CheckInterrupt(f)
+
+  /**
+   * Makes an explicit check to see if the fiber has been interrupted, and if
+   * so, performs self-interruption.
+   */
+  final def allowInterrupt: UIO[Unit] =
+    descriptor.flatMap(d => if (d.interrupted) interrupt else unit)
+
+  /**
+   * Makes the effect uninterruptible, but passes it a restore function that
+   * can be used to restore the inherited interruptibility from whatever region
+   * the effect is composed into.
+   */
+  final def uninterruptibleMask[R >: LowerR, E <: UpperE, A](
+    f: ZIO.InterruptStatusRestore => ZIO[R, E, A]
+  ): ZIO[R, E, A] =
+    checkInterruptible(flag => f(new ZIO.InterruptStatusRestore(flag)).uninterruptible)
```

Comment:
Now should we be mindful of marking all ZIO primitives as interruptible when they enter a blocking state? So that they remain so, even if marked as `uninterruptible` by the user?

Reply:
I think so. That's basically what Haskell does. At the very least, we should do it for any "blocking" operations, like sleeping.